Prevent the bot from replying multiple text messages

firelightning13 opened the pull request:
Currently, the bot will reply to multiple people all at once, which could be unintended effect, especially when you have multiple people in the text channel. (assuming ALWAYS_REPLY is set to true)

For example:
![image](https://github.com/ausboss/PygDiscordBot/assets/17120382/13c7de37-c063-45c0-97ad-b05c08f3de99)

This PR "should" fix this problem, where the bot will only reply once. The current AI model nowadays are pretty great at summarize what happened in the chat room (evident by just using listen-only mode).

The intended way for bot to reply (note that I sent replies to the bot 2 times instead of one):
![image](https://github.com/ausboss/PygDiscordBot/assets/17120382/7fb6f999-30b9-4364-920e-99873bcb2cb3)

### To-do checklist
- [x] Implement task creation and cancellation from asyncio library
- [x] Use aiohttp to create POST session instead of usual POST request to prevent blocking
- [x] Resolved memory issues (bot can't see 2nd last reply)
- [x] TextGen compatible (though it doesn't prevent server from stopping the generation because of API limitation)
- [x] KoboldAI API compatible
- [x] Koboldcpp compatible (generate unique key to API to support multiple channel ID)
- [x] Multiple channel ID support

**Currently, this PR is untested for multiple channel IDs, so if you want to test this implementation, you can pull my main repo instead.**

firelightning13 committed:
Prevent the bot from replying multiple replies
- Implement asynchronous call in textgen.py
- Use asyncio cancel the current task and start a new one each time a text message is received

firelightning13 commented:
If you want to know the current problem is, I made a simple test:

![image](https://github.com/ausboss/PygDiscordBot/assets/17120382/d7feda16-d046-47b2-9d7f-7da7b3120d71)

In the logs, notice the first sentence is missing in `memory`:

Neuro Sama: {'input': 'FL13: osu?', 'stop': ['\n\n', '<END>', 'You', '\nFL13:'], 'history': "FL13: hello again, neuro sama\nNeuro sama: Hello again! Nice to see you back in the chat. Are you ready for some more fun conversations?\nFL13: yeah, how are you?\nNeuro sama: I'm doing fantastic, thank you for asking! Always happy to chat with everyone in the server. 😊", 'response': " Osu! You mean the popular rhythm game? Absolutely! It's one of my personal favorites. The fast-paced action and catchy tunes never fail to get my adrenaline pumping. But don't worry, I also enjoy playing other genres too. Variety keeps life interesting, right?\n"}

firelightning13 committed:
Fix bot memory 2nd last user reply issue
- Bot can now see 2nd last user reply by storing last message before send another reply
- Avoid `NoneType` error in messagehandler.py, I forgor

firelightning13 committed:
Fix stop sequence for TextGen in `_acall` function

- I literally forgot

firelightning13 committed:
Made KoboldAI API compatible
- implementation carried over to koboldAI API
- so far tested only with koboldcpp

ausboss commented:
👀I like what I see. I am going to play around with this more and possibly merge soon.

firelightning13 committed:
Stop generate text via API (koboldcpp only)
- Made a `_stop` function in koboldai.py
- By invoking `/api/extra/abort`
- Official KoboldAI doesn't work

firelightning13 commented:
> 👀I like what I see. I am going to play around with this more and possibly merge soon.

@ausboss Thanks for the reply. I only had a short time testing it, but so far there's no problem in my end. However, I do notice that it supports multiple channel IDs, so this may pose some problems. (it will likely cancel the reply from other channel)

So when cancelling the task, it should check the channel ID as well. I'll look into it later for the solution, but I'm not a full time programmer though. (I'm a graphic designer lol)

firelightning13 commented:
I think I know what you meant by that. Maybe if I use `self.current_task` as a dictionary and then go through a loop that would check the `channel_id`, this could potentially solve the problem. I'll see if asyncio would handle this sort of thing, and figure out how to clear any tasks that has been `done`.

FYI, I actually solved the multiple message issue a long time ago, but I didn't make a PR (you can look at the date in the screenshot haha) because I heavily modified your code. Back then I changed it to OpenAI API compatible which defeats the purpose of this project,  and that was before you implemented Langchain. If I were to implement now, I have to change the message format into chat completion object (which is pain in the ass). So, I don't even bother to do it anymore.

firelightning13 committed:
Support abrupt replies from multiple channel IDs
- Using dict for current_tasks, with channel IDs as a key
- Perform a check if the endpoint is using koboldcpp or KoboldAI
- TODO: support koboldcpp with multi-user requests

firelightning13 committed:
Implement `genkey` to support multiple channel IDs (NOT TESTED!)
- using dict to generate unique keys for each of channel IDs
- `_stop` function rolled back to normal request POST instead of aiohttp session

firelightning13 committed:
Oops

firelightning13 committed:
Merge branch 'ausboss:main' into main
